source/extensions/filters/http/cdn_loop/parser.cc

```diff
@@ -0,0 +1,359 @@
+// This rule allows lower case letters too in violation of the RFC.
+constexpr bool isHexDigitCaseInsensitive(char c) {
+  return isDigit(c) || ('A' <= c && c <= 'F') || ('a' <= c && c <= 'f');
+}
+
+// RFC 7230 Section 3.2.6 defines obs-text as:
+//
+// obs-text       = %x80-FF
+constexpr bool isObsText(char c) { return 0x80 & c; }
+
+// RFC 7230 Section 3.2.6 defines qdtext as:
+//
+// qdtext         = HTAB / SP / %x21 / %x23-5B / %x5D-7E / obs-text
+constexpr bool isQdText(char c) {
+  return c == '\t' || c == ' ' || c == '\x21' || ('\x23' <= c && c <= '\x5B') ||
+         ('\x5D' <= c && c <= '\x7E') || isObsText(c);
+}
+
+// RFC 5324 Appendix B.1 says:
+//
+// VCHAR          =  %x21-7E
+//                        ; visible (printing) characters
+constexpr bool isVChar(char c) { return '\x21' <= c && c <= '\x7e'; }
+
+} // namespace
+
+ParseContext parseOptionalWhitespace(const ParseContext& input) {
+  ParseContext context = input;
+  while (!context.atEnd()) {
+    const char c = context.peek();
+    if (!(c == ' ' || c == '\t')) {
+      break;
+    }
+    context.increment();
+  }
+  return context;
+}
+
+StatusOr<ParseContext> parseQuotedPair(const ParseContext& input) {
+  ParseContext context = input;
+  if (context.atEnd()) {
+    return absl::InvalidArgumentError(
+        absl::StrFormat("expected backslash at position %d; found end-of-input", context.next()));
+  }
+
+  if (context.peek() != '\\') {
+    return absl::InvalidArgumentError(absl::StrFormat(
+        "expected backslash at position %d; found '%c'.", input.next(), context.peek()));
+  }
+  context.increment();
+
+  if (context.atEnd()) {
+    return absl::InvalidArgumentError(absl::StrFormat(
+        "expected escaped character at position %d; found end-of-input", context.next()));
+  }
+
+  const char c = context.peek();
+  if (!(c == '\t' || c == ' ' || isVChar(c) || isObsText(c))) {
+    return absl::InvalidArgumentError(
+        absl::StrFormat("expected escapable character at position %d; found '\\x%x'.", input.next(),
+                        context.peek()));
+  }
+  context.increment();
+
+  return context;
+}
+
+StatusOr<ParseContext> parseQuotedString(const ParseContext& input) {
+  ParseContext context = input;
+
+  if (context.atEnd()) {
+    return absl::InvalidArgumentError(absl::StrFormat(
+        "expected opening '\"' at position %d; found end-of-context", context.next()));
+  }
+
+  if (context.peek() != '"') {
+    return absl::InvalidArgumentError(absl::StrFormat(
+        "expected opening quote at position %d; found '%c'.", context.next(), context.peek()));
+  }
+  context.increment();
+
+  while (!context.atEnd() && context.peek() != '"') {
+    if (isQdText(context.peek())) {
+      context.increment();
+      continue;
+    }
+
+    if (StatusOr<ParseContext> quoted_pair_context = parseQuotedPair(context);
+        !quoted_pair_context) {
+      return quoted_pair_context.status();
+    } else {
+      context.setNext(*quoted_pair_context);
```

Comment:
All of the parser functions call either `increment` or `setNext` on a context. (In some cases, like parseOptionalWhitespace, a particular invocation may not call either `increment` or `setNext`.)

The problem with passing in the context as a non-const is that undoing the changes is harder since we'd have to wind back any call to `increment` or `setNext`.  Contexts are pretty small (24 bytes on a 64-bit platform), so having a local copy seemed like a good idea to me.